src/task/task.cpp

```diff
@@ -1,51 +1,142 @@
+                 bool should_skip_query_cache, util::TupleFunction &&tuple_fn)
+    : Task(db_oid, std::move(query_text), policy, sync),
+      cost_model_(std::move(cost_model)),
+      params_(std::move(params)),
+      param_types_(std::move(param_types)),
+      settings_(settings),
+      override_qid_(override_qid),
+      metrics_manager_(metrics_manager),
+      should_force_abort_(should_force_abort),
+      should_skip_query_cache_(should_skip_query_cache),
+      tuple_fn_(std::move(tuple_fn)) {}
+
 void TaskDDL::Execute(common::ManagedPointer<util::QueryExecUtil> query_exec_util,
                       common::ManagedPointer<task::TaskManager> task_manager) {
-  query_exec_util->BeginTransaction(db_oid_);
-  bool status = query_exec_util->ExecuteDDL(query_text_, false);
-  query_exec_util->EndTransaction(status);
-
-  if (sync_) {
-    if (status)
-      sync_->Success(DummyResult{});
-    else
+  query_exec_util->BeginTransaction(db_oid_, policy_);
+  bool ddl_ok = query_exec_util->ExecuteDDL(query_text_, false);
+  query_exec_util->EndTransaction(ddl_ok);
+
+  if (sync_ != DISABLED) {
+    if (ddl_ok) {
+      sync_->Success(common::FutureDummyResult{});
+    } else {
       sync_->Fail(query_exec_util->GetError());
+    }
   }
 }
 
 void TaskDML::Execute(common::ManagedPointer<util::QueryExecUtil> query_exec_util,
                       common::ManagedPointer<task::TaskManager> task_manager) {
-  bool result = true;
-  query_exec_util->BeginTransaction(db_oid_);
+  bool dml_ok = true;  // Track whether the DML was executed successfully.
+  query_exec_util->BeginTransaction(db_oid_, policy_);
 
-  if (skip_query_cache_) {
+  if (should_skip_query_cache_) {
     // Skipping the query cache is implemented by invalidating the plan
     // and then invalidating the plan entry after execution.
     query_exec_util->ClearPlan(query_text_);
   }
 
   if (params_.empty()) {
-    result = query_exec_util->ExecuteDML(query_text_, nullptr, nullptr, tuple_fn_, metrics_manager_,
-                                         std::make_unique<optimizer::TrivialCostModel>(), override_qid_, settings_);
+    dml_ok = query_exec_util->ExecuteDML(query_text_, nullptr, nullptr, tuple_fn_, metrics_manager_,
+                                         std::move(cost_model_), override_qid_, settings_);
   } else {
     std::vector<parser::ConstantValueExpression> &params_0 = params_[0];
-    result = query_exec_util->CompileQuery(query_text_, common::ManagedPointer(&params_0),
+    dml_ok = query_exec_util->CompileQuery(query_text_, common::ManagedPointer(&params_0),
                                            common::ManagedPointer(&param_types_), std::move(cost_model_), override_qid_,
                                            settings_);
 
     // Execute with specified parameters only if compilation succeeded
-    if (result) {
+    if (dml_ok) {
       for (auto &param_vec : params_) {
```

Comment:
We need some new construct that is a combination of both `for`-loop and `while`-loop...

Reply:
specifically, range-based for and while loop, you could have stuffed dml_ok into the cond